### PR TITLE
[LETS-55] parallel recovery redo unit test

### PR DESCRIPTION
**Pull request by cristiarg**
http://jira.cubrid.org/browse/LETS-55

A unit test to verify correctness of the infrastructure used to apply log recovery redo in parallel.

There are two sets of usecases:
- marked with `[ci]` are supposed to execute quickly (ci = continuous integration)
- marked with `[long]` are of the "stress-test" type.

The `[ci]` tests have also been added to the `run_scalability_unit_tests.sh` to be by the CI infrastructure.

Implementation:
- a custom unit-test specific implementation of `cublog::redo_parallel::redo_job_base` has been added - `ut_redo_job_impl`
- that leverages a mock database (`ut_database`) / volume (`ut_volume`) / pages (`ut_page`) object structure that generates and then applies actions (jobs) akin to the log_rec's that are present in the real database log
- upon generating such (jobs) an initial ("online") instance of the mock database is constructed
- upon applying the changes (done by the - under test - parallel infrastructure) another ("recovery") instance of the mock database is constructed
- in the en the "online and the "recovery" structures must match verbatim

### CODE REVIEW

File: unit_tests/recovery/ut_database.cpp [hunk 1]
```diff
@@ -0,0 +1,363 @@
...
+  : m_database_config (a_database_config)
+  , m_lsa_log_id (10100) // just starts from an arbitrary value
+  , m_gen (m_rd ())
+  , m_duration_in_millis_dist (1, a_database_config.max_duration_in_millis)
+  // *INDENT-OFF*
+  , m_add_or_update_volume_dist ({ut_database_config::ADD_VOLUME_DISCRETE_RATIO,
+                                 ut_database_config::UPDATE_VOLUME_DISCRETE_RATIO })
+  , m_add_or_update_page_dist ({ut_database_config::ADD_PAGE_DISCRETE_RATIO,
+                               ut_database_config::UPDATE_PAGE_DISCRETE_RATIO})
+  // *INDENT-ON*
```
Comment (ribram) [inline unit_tests/recovery/ut_database.cpp:35]: Being a cpp file, indent does not run on it, so the comments have no effect.
```suggestion
  , m_add_or_update_volume_dist ({ut_database_config::ADD_VOLUME_DISCRETE_RATIO,
                                 ut_database_config::UPDATE_VOLUME_DISCRETE_RATIO })
  , m_add_or_update_page_dist ({ut_database_config::ADD_PAGE_DISCRETE_RATIO,
                               ut_database_config::UPDATE_PAGE_DISCRETE_RATIO})
```

File: unit_tests/recovery/ut_database.cpp [hunk 2]
```diff
@@ -0,0 +1,363 @@
...
+	  || m_pages.size () == 0)
+	{
+	  // add new page and generate log entry
+	  // akin to 'extend volume' operation
+	  const auto lsa_log_id = a_db_global_values.increment_and_get_lsa_log_id ();
+	  const int millis = a_db_global_values.rand_duration_in_millis ();
+	  ux_ut_redo_job_impl job
+	  {
+	    new ut_redo_job_impl (a_database_recovery, ut_redo_job_impl::job_type::NEW_PAGE,
+	    lsa_log_id, {static_cast<short> (m_pages.size ()), m_volid}, millis)
```
Comment (ribram) [inline unit_tests/recovery/ut_database.cpp:203]: Why is pageid cast to short?

File: unit_tests/recovery/ut_database.cpp [hunk 3]
```diff
@@ -0,0 +1,363 @@
...
+  : m_database_config (a_database_config)
+{
+  // start without any volumes
+}
+
+ux_ut_redo_job_impl ut_database::generate_changes (ut_database &a_database_recovery,
+    ut_database_values_generator &a_db_global_values)
+{
+  //const int update_or_add_volume = rand_update_or_add_volume();
+  const auto add_or_update_volume = a_db_global_values.rand_add_or_update_volume (m_volumes.size ());
```
Comment (ribram) [inline unit_tests/recovery/ut_database.cpp:284]: A suggestion to avoid checking m_volumes.size() again.
```suggestion
  auto add_or_update_volume = ut_database_values_generator::add_or_update::ADD;
  if (m_volumes.size () > 0)
    {
      add_or_update_volume = a_db_global_values.rand_add_or_update_volume (m_volumes.size ());
    }
```

File: run_scalability_unit_tests.sh [hunk 4]
```diff
@@ -20,3 +20,4 @@
 ./test_prior_list_serialize
 ./test_prior_sendrecv
 ./test_request_cs
+./test_recovery "[ci]"
```
[the remark this answers is not in the document]
Comment (cristiarg) [reply run_scalability_unit_tests.sh:23]: Done

File: unit_tests/CMakeLists.txt [hunk 5]
```diff
@@ -37,6 +37,7 @@ option (UNIT_TEST_MONITOR "Unit testing: monitor")
 option (UNIT_TEST_LOADDB "Unit testing: loaddb module")
 option (UNIT_TEST_REQUEST_CS "Unit testing: request_client_server module")
 option (UNIT_TEST_LOG_TRANSFER "Unit testing: log transfer")
+option (UNIT_TEST_LOG_RECOVERY_PARALLEL_POC "Unit testing: log recovery parallel poc")
```
[the remark this answers is not in the document]
Comment (cristiarg) [reply unit_tests/CMakeLists.txt:40]: Nothing against. But let's do in another issue/PR. Add an issue and assign to me.

File: unit_tests/recovery/CMakeLists.txt [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+#  GNU General Public License for more details.
+#
+#  You should have received a copy of the GNU General Public License
+#  along with this program; if not, write to the Free Software
+#  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
+#
+
+project (test_recovery)
+
+set (RECOVERY_SRC
```
[the remark this answers is not in the document]
Comment (cristiarg) [reply unit_tests/recovery/CMakeLists.txt:20]: Done

File: unit_tests/recovery/CMakeLists.txt [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+  ${TRANSACTION_DIR}/log_recovery_redo_parallel.cpp
+  ut_redo_job_impl.cpp
+  ut_database.cpp
+  main.cpp
+  )
+
+set (RECOVERY_H
```
[the remark this answers is not in the document]
Comment (cristiarg) [reply unit_tests/recovery/CMakeLists.txt:27]: Done

File: unit_tests/recovery/main.cpp [hunk 8]
```diff
@@ -0,0 +1,208 @@
...
+  /* the number of async task for the algorithm to use internally */
+  const size_t parallel_count;
+
+  /* the number of redo jobs to generate */
+  const size_t redo_job_count;
+
+  bool verbose;
+};
+
+static bool thread_infrastructure_initialized = bool ();
```
[the remark this answers is not in the document]
Comment (cristiarg) [reply unit_tests/recovery/main.cpp:35]: It is.
As per https://stackoverflow.com/questions/17690095/c-initialization-of-int-variables-by-an-implicit-constructor (draft at : https://isocpp.org/std/the-standard )

File: unit_tests/recovery/main.cpp [hunk 9]
```diff
@@ -0,0 +1,208 @@
...
+
+//      if (a_test_config.verbose)
+//        {
+//          std::cout << "P: "
+//                    << "syn_" << ((log_entry->get_is_to_be_waited_for_op ()) ? '1' : '0')
+//                    << std::setw (4) << log_entry->get_vpid ().volid << std::setfill ('_')
+//                    << std::setw (5) << log_entry->get_vpid ().pageid << std::setfill (' ')
+//                    << "  eids: " << std::setw (7) << log_entry->get_entry_id ()
+//                    << std::endl;
+//        }
```
[the remark this answers is not in the document]
Comment (cristiarg) [reply unit_tests/recovery/main.cpp:107]: Done

File: unit_tests/recovery/main.cpp [hunk 10]
```diff
@@ -0,0 +1,208 @@
...
+
+  cublog::redo_parallel log_redo_parallel (a_test_config.parallel_count);
+
+  ux_ut_database db_online { new ut_database (a_database_config) };
+
+  ux_ut_database db_recovery { new ut_database (a_database_config) };
+
+#if (0)
+  consumption_accumulator dbg_accumulator;
+#endif
```
[the remark this answers is not in the document]
Comment (cristiarg) [reply unit_tests/recovery/main.cpp:78]: Done

File: unit_tests/recovery/main.cpp [hunk 11]
```diff
@@ -0,0 +1,208 @@
...
+{
+  if (a_test_config.verbose)
+    {
+      std::cout << "TEST:"
+		<< "\t volumes=" << a_database_config.max_volume_count_per_database
+		<< "\t pages=" << a_database_config.max_page_count_per_volume
+		<< "\t parallel=" << a_test_config.parallel_count
+		<< "\t jobs=" << a_test_config.redo_job_count
+		<< std::endl;
+    }
```
[the remark this answers is not in the document]
Comment (cristiarg) [reply unit_tests/recovery/main.cpp:67]: Just to see progress when debugging.

File: unit_tests/recovery/main.cpp [hunk 12]
```diff
@@ -0,0 +1,208 @@
...
+//                    << std::setw (5) << log_entry->get_vpid ().pageid << std::setfill (' ')
+//                    << "  eids: " << std::setw (7) << log_entry->get_entry_id ()
+//                    << std::endl;
+//        }
+    }
+
+  log_redo_parallel.set_adding_finished ();
+  log_redo_parallel.wait_for_termination_and_stop_execution ();
+
+  return *db_online == *db_recovery;
```
[the remark this answers is not in the document]
Comment (cristiarg) [reply unit_tests/recovery/main.cpp:113]: Done

File: unit_tests/recovery/main.cpp [hunk 13]
```diff
@@ -0,0 +1,208 @@
...
+	    const ut_database_config database_config =
+	    {
+	      volume_count_per_database, // max_volume_count_per_database
+	      page_count_per_volume, // max_page_count_per_volume
+	      0, // max_duration_in_millis
+	    };
+
+	    const bool test_result = execute_test (test_config, database_config);
+	    REQUIRE (test_result);
+	  }
```
[the remark this answers is not in the document]
Comment (cristiarg) [reply unit_tests/recovery/main.cpp:174]: Done

File: unit_tests/recovery/main.cpp [hunk 14]
```diff
@@ -0,0 +1,208 @@
...
+    false, // verbose
+  };
+
+  cublog::redo_parallel log_redo_parallel (test_config.parallel_count);
+
+  log_redo_parallel.set_adding_finished ();
+  log_redo_parallel.wait_for_termination_and_stop_execution ();
+
+  REQUIRE (true);
+}
```
[the remark this answers is not in the document]
Comment (cristiarg) [reply unit_tests/recovery/main.cpp:142]: The test is relevant for the empty 'corner case' when there's nothing to execute and the code should not hang/crash/... . So I say the test is relevant. As to the form to test it ... I'm open to improvements.

File: unit_tests/recovery/main.cpp [hunk 15]
```diff
@@ -0,0 +1,208 @@
...
+
+TEST_CASE ("log recovery parallel test 2: some jobs, some tasks", "[ci][dbg]")
+{
+  srand (time (nullptr));
+  initialize_thread_infrastructure ();
+
+  std::array<size_t, 3> volume_count_per_database_arr { 1u, 2u, 10u };
+  std::array<size_t, 2> page_count_per_volume_arr { 10u, _1k };
```
[the remark this answers is not in the document]
Comment (cristiarg) [reply unit_tests/recovery/main.cpp:150]: It's just one of the testcases. The combination of few/many pages/volumes is designed to set different contention levels on a resource.

File: unit_tests/recovery/ut_database.cpp (hunk 1 above)
Comment (cristiarg) [reply unit_tests/recovery/ut_database.cpp:35]: I found that it does. Maybe `astyle` also honors them.

File: unit_tests/recovery/main.cpp [hunk 16]
```diff
@@ -0,0 +1,208 @@
...
+
+  cublog::redo_parallel log_redo_parallel (test_config.parallel_count);
+
+  log_redo_parallel.set_adding_finished ();
+  log_redo_parallel.wait_for_termination_and_stop_execution ();
+
+  REQUIRE (true);
+}
+
+TEST_CASE ("log recovery parallel test 2: some jobs, some tasks", "[ci][dbg]")
```
[the remark this answers is not in the document]
Comment (cristiarg) [reply unit_tests/recovery/main.cpp:144]: close to 30 mins

File: unit_tests/recovery/main.cpp [hunk 17]
```diff
@@ -0,0 +1,208 @@
...
+	  }
+}
+
+TEST_CASE ("log recovery parallel test N: stress test", "[long]")
+{
+  srand (time (nullptr));
+  initialize_thread_infrastructure ();
+
+  std::array<size_t, 3> volume_count_per_database_arr { 1u, 2u, 10u };
+  std::array<size_t, 3> page_count_per_volume_arr { 10u, _1k, _16k };
```
[the remark this answers is not in the document]
Comment (cristiarg) [reply unit_tests/recovery/main.cpp:183]: So maybe the guiding idea is to use a small value and a large value for the volume/page count.

File: unit_tests/recovery/ut_database.cpp (hunk 2 above)
Comment (cristiarg) [reply unit_tests/recovery/ut_database.cpp:203]: Fixed

File: unit_tests/recovery/ut_database.cpp (hunk 3 above)
Comment (cristiarg) [reply unit_tests/recovery/ut_database.cpp]: Done

File: unit_tests/CMakeLists.txt (hunk 5 above)
[the remark this answers is not in the document]
Comment (ribram) [reply unit_tests/CMakeLists.txt:40]: Agreed

File: unit_tests/recovery/main.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (ribram) [reply unit_tests/recovery/main.cpp:35]: Sorry, I mean isn't it better to initialize as false? I guess `bool()` is also false, but the intention is more clear when the specific value is used.

File: unit_tests/recovery/main.cpp (hunk 11 above)
[the remark this answers is not in the document]
Comment (ribram) [reply unit_tests/recovery/main.cpp:67]: ok

File: unit_tests/recovery/main.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (cristiarg) [reply unit_tests/recovery/main.cpp]: Done. It is indeed more explicit to the next developer and also wiser to not "stretch" the conformance of the compiler.

File: unit_tests/recovery/main.cpp (hunk 16 above)
[the remark this answers is not in the document]
Comment (ribram) [reply unit_tests/recovery/main.cpp]: Let's make a rule to keep every unit test under 1 second.

 My suggestion for the CI version of this unit test is to use a only two executions - one with a single thread and one with the hardware_concurrency - with a single volume/page/job count configuration (e.g. 10 volumes, 1k pages, 1k jobs). And let the long version of the test run all kinds of variations.

[the remark this answers is not in the document]
Comment (cristiarg) [reply unit_tests/recovery/main.cpp]: Done. However I put the number of jobs at 64k where execution time is ~0.5 seconds.

File: unit_tests/recovery/main.cpp [hunk 18]
```diff
@@ -0,0 +1,190 @@
...
+
+	    const ut_database_config database_config =
+	    {
+	      volume_count_per_database, // max_volume_count_per_database
+	      page_count_per_volume, // max_page_count_per_volume
+	      0., // max_duration_in_millis
+	    };
+
+	    execute_test (test_config, database_config);
+	  }
```
Comment (ribram) [inline unit_tests/recovery/main.cpp:148]: Add the scopes for each for loop, similar to the long test case.

**Review by ribram — APPROVED** [review]
Please add the scopes for the for loops and merge the pull request

File: unit_tests/recovery/main.cpp (hunk 18 above)
Comment (cristiarg) [reply unit_tests/recovery/main.cpp]: Done
